cpp/src/arrow/engine/substrait/relation_internal.cc

```diff
@@ -91,6 +89,106 @@ Result<DeclarationInfo> ProcessEmit(const RelMessage& rel,
   }
 }
 
+template <typename RelMessage>
+Result<DeclarationInfo> ProcessEmit(const RelMessage& rel,
+                                    const DeclarationInfo& no_emit_declr,
+                                    const std::shared_ptr<Schema>& schema) {
+  return ProcessEmit(rel.has_common() ? std::optional(rel.common()) : std::nullopt,
+                     no_emit_declr, schema);
+}
+
+Result<DeclarationInfo> ProcessExtensionEmit(
+    const DeclarationInfo& no_emit_declr, const std::vector<int>& emit_info,
+    const std::vector<int>& field_output_indices) {
+  const std::shared_ptr<Schema>& input_schema = no_emit_declr.output_schema;
+  std::vector<compute::Expression> proj_field_refs;
+  proj_field_refs.reserve(emit_info.size());
+  FieldVector emit_fields;
+  emit_fields.reserve(emit_info.size());
+
+  for (int emit_idx : emit_info) {
+    if (emit_idx < 0 || static_cast<size_t>(emit_idx) >= field_output_indices.size()) {
+      return Status::Invalid("Out of bounds emit index ", emit_idx);
+    }
+    int field_idx = field_output_indices[emit_idx];
+    if (field_idx < 0) {
+      return Status::Invalid("Non-output emit index ", emit_idx);
+    }
+    proj_field_refs.push_back(compute::field_ref(FieldRef(field_idx)));
+    emit_fields.push_back(input_schema->field(field_idx));
+  }
+
+  std::shared_ptr<Schema> emit_schema = schema(std::move(emit_fields));
+
+  return DeclarationInfo{
+      compute::Declaration::Sequence(
+          {no_emit_declr.declaration,
+           {"project", compute::ProjectNodeOptions{std::move(proj_field_refs)}}}),
+      std::move(emit_schema)};
+}
+
+Result<RelationInfo> GetExtensionRelationInfo(const substrait::Rel& rel,
+                                              const ExtensionSet& ext_set,
+                                              const ConversionOptions& conv_opts,
+                                              std::vector<DeclarationInfo>* inputs_arg) {
```

Comment:
What if one of the input_args is another "asof join"? Do we care / need to pass the `field_output_indices` of the input asof join relation to this function?

Reply:
I don't see `field_output_indices` here. Do you mean "pass `input_args`"? This `input_args` is an output parameter and `GetExtensionRelationInfo` is an internal function, so I'm not sure what you have in mind when thinking about a second asof-join.

Reply:
The case I am thinking is whether `field_output_indices` needs to be passed across different asof join node or not

Reply:
No, `field_output_indices` need not be passed. It is a property of the relation that is not affected by other relations in the plan.